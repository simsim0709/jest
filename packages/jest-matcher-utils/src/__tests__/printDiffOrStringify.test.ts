/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {alignedAnsiStyleSerializer} from '@jest/test-utils';
import {INVERTED_COLOR, printDiffOrStringify} from '../index';

expect.addSnapshotSerializer(alignedAnsiStyleSerializer);

describe('printDiffOrStringify', () => {
  const testDiffOrStringify = (expected: string, received: string): string =>
    printDiffOrStringify(expected, received, 'Expected', 'Received', true);

  test('expected is empty and received is single line', () => {
    const expected = '';
    const received = 'single line';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('expected is multi line and received is empty', () => {
    const expected = 'multi\nline';
    const received = '';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('expected and received are single line with multiple changes', () => {
    const expected = 'delete common expected common prev';
    const received = 'insert common received common next';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('expected and received are multi line with trailing spaces', () => {
    const expected = 'delete \ncommon expected common\nprev ';
    const received = 'insert \ncommon received common\nnext ';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('has no common after clean up chaff multiline', () => {
    const expected = 'delete\ntwo';
    const received = 'insert\n2';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  test('has no common after clean up chaff one-line', () => {
    const expected = 'delete';
    const received = 'insert';
    expect(testDiffOrStringify(expected, received)).toMatchSnapshot();
  });

  describe('MAX_DIFF_STRING_LENGTH', () => {
    const lessChange = INVERTED_COLOR('single ');
    const less = 'single line';
    const more = 'multi line' + '\n123456789'.repeat(2000); // 10 + 20K chars

    test('both are less', () => {
      const difference = testDiffOrStringify('multi\nline', less);

      expect(difference).toMatch('- multi');
      expect(difference).toMatch('- line');

      // diffStringsUnified has substring change
      expect(difference).not.toMatch('+ single line');
      expect(difference).toMatch(lessChange);
    });

    test('expected is more', () => {
      const difference = testDiffOrStringify(more, less);

      expect(difference).toMatch('- multi line');
      expect(difference).toMatch('+ single line');

      // diffLinesUnified does not have substring change
      expect(difference).not.toMatch(lessChange);
    });

    test('received is more', () => {
      const difference = testDiffOrStringify(less, more);

      expect(difference).toMatch('- single line');
      expect(difference).toMatch('+ multi line');

      // diffLinesUnified does not have substring change
      expect(difference).not.toMatch(lessChange);
    });
  });
});
