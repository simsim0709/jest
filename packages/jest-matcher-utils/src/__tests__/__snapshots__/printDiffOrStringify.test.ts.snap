// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printDiffOrStringify expected and received are multi line with trailing spaces 1`] = `
<g>- Expected  - 3</>
<r>+ Received  + 3</>

<g>- <i>delete</i><Y> </></>
<r>+ <i>insert</i><Y> </></>
<g>- common <i>expect</i>ed common</>
<r>+ common <i>receiv</i>ed common</>
<g>- <i>prev</i><Y> </></>
<r>+ <i>next</i><Y> </></>
`;

exports[`printDiffOrStringify expected and received are single line with multiple changes 1`] = `
Expected: <g>"<i>delete</i> common <i>expect</i>ed common <i>prev</i>"</>
Received: <r>"<i>insert</i> common <i>receiv</i>ed common <i>next</i>"</>
`;

exports[`printDiffOrStringify expected is empty and received is single line 1`] = `
Expected: <g>""</>
Received: <r>"single line"</>
`;

exports[`printDiffOrStringify expected is multi line and received is empty 1`] = `
Expected: <g>"multi</>
<g>line"</>
Received: <r>""</>
`;

exports[`printDiffOrStringify has no common after clean up chaff multiline 1`] = `
<g>- Expected  - 2</>
<r>+ Received  + 2</>

<g>- delete</>
<g>- two</>
<r>+ insert</>
<r>+ 2</>
`;

exports[`printDiffOrStringify has no common after clean up chaff one-line 1`] = `
Expected: <g>"delete"</>
Received: <r>"insert"</>
`;
